packages/nextjs/components/scaffold-eth/Contract/ContractUI.tsx

```diff
@@ -1,82 +1,194 @@
-import { useReducer } from "react";
+import { useEffect, useReducer, useState } from "react";
+import router from "next/router";
 import { ContractReadMethods } from "./ContractReadMethods";
 import { ContractVariables } from "./ContractVariables";
 import { ContractWriteMethods } from "./ContractWriteMethods";
 import { Abi } from "viem";
-import { Address, Balance } from "~~/components/scaffold-eth";
+import { HeartIcon } from "@heroicons/react/24/outline";
+import { BuidlGuidlLogo } from "~~/components/assets/BuidlGuidlLogo";
+import { Address, Balance, MethodSelector } from "~~/components/scaffold-eth";
 import { useNetworkColor } from "~~/hooks/scaffold-eth";
 import { useAbiNinjaState } from "~~/services/store/store";
 import { getTargetNetworks } from "~~/utils/scaffold-eth";
 
+type ExtractAbiFunction<T> = T extends { type: "function" } ? T : never; // @todo duplicate in MethodSelector
+type AbiFunction = ExtractAbiFunction<Abi[number]>;
+
 type ContractUIProps = {
   className?: string;
-  deployedContractData: { address: string; abi: Abi };
+  initialContractData: { address: string; abi: Abi };
 };
 
 const mainNetworks = getTargetNetworks();
 
 /**
  * UI component to interface with deployed contracts.
  **/
-export const ContractUI = ({ className = "", deployedContractData }: ContractUIProps) => {
+export const ContractUI = ({ className = "", initialContractData }: ContractUIProps) => {
   const [refreshDisplayVariables, triggerRefreshDisplayVariables] = useReducer(value => !value, false);
   const mainChainId = useAbiNinjaState(state => state.mainChainId);
   const mainNetwork = mainNetworks.find(network => network.id === mainChainId);
   const networkColor = useNetworkColor(mainNetwork);
 
+  const updateUrlWithSelectedMethods = (selectedMethods: string[]) => {
+    const currentQuery = new URLSearchParams(window.location.search);
+    if (selectedMethods.length > 0) {
+      currentQuery.set("methods", selectedMethods.join(","));
+    } else {
+      currentQuery.delete("methods");
+    }
+    const newPath = `/${initialContractData.address}/${mainChainId}`;
+
+    router.push({ pathname: newPath, query: currentQuery.toString() }, undefined, { shallow: true });
+  };
+
+  // include all functions with inputs
+  const methodsWithInputs = initialContractData.abi.filter(
```

Comment:
I think change it to something like this : 

```ts
  const readMethodsWithInputsAndWriteMethods = initialContractData.abi.filter((method): method is AbiFunction => {
    if (method.type !== "function") return false;

    // Check for read functions
    if (method.stateMutability === "view" || method.stateMutability === "pure") {
      // Check for read inputs length
      if (method.inputs.length > 0) {
        return true;
      } else return false;
    } else {
      // Else condition defines write methods
      return true;
    }
  });
```

Since we also want to show write methods with 0 inputs in `MethodSelector` we should also probably rename `MethodSelector` prop

PS: Sorry for the long variable name 😅 please feel free to use any other also point out any mistake in logic 🙌

Reply:
Thanks this is much more understandable!